### ISSUE TITLE
div(curl) with H(curl, G) is non-zero depending on parallel or serial refinement

### ISSUE THREAD

**Issue by kvoronin** [mfem/mfem#419]
Hello,

I've recently faced the following when I was trying to check the div(curl) = 0 equality at the matrix level.
The check is that I assemble divergence matrix from H(div) to L2, and Curl operator from H(curl,G) to Hdiv where I have some essential boundary dofs eliminated via EliminateTrialDofs. Then I do ParMult for HypreParMatrices and check that both diagonal and off-diagonal blocks are zero.

When a Mesh is first refined in serial and then a ParMesh is created, I am getting 0 as it should be.
When first a ParMesh is created and only then UniformRefinement is called, I am getting for nproc > 4 nonzero norm of the off-diagonal part for some of the processes.

Can someone please give an insight of what is happening and how to avoid that?
If the whole boundary in Hcurl is marked as essential, nothing bad happened. So I assume something is wrong with extracting boundary dofs in H(curl).

My MFEM version is based on MFEM 3.2 (so if it doesn't reproduce in mfem 3-3 let me know).

Code is below:
```
//                       MFEM Example 1 - Parallel Version modified

#include "mfem.hpp"
#include <fstream>
#include <iostream>

using namespace std;
using namespace mfem;

int main(int argc, char *argv[])
{
   // 1. Initialize MPI.
   int num_procs, myid;
   MPI_Init(&argc, &argv);
   MPI_Comm comm = MPI_COMM_WORLD;
   MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
   MPI_Comm_rank(MPI_COMM_WORLD, &myid);

   // 2. Parse command-line options.
   const char *mesh_file = "../data/cube_3d_moderate.mesh";
   int order = 1;
   bool static_cond = false;
   bool visualization = 1;

   OptionsParser args(argc, argv);
   args.AddOption(&mesh_file, "-m", "--mesh",
                  "Mesh file to use.");
   args.AddOption(&order, "-o", "--order",
                  "Finite element order (polynomial degree) or -1 for"
                  " isoparametric space.");
   args.AddOption(&static_cond, "-sc", "--static-condensation", "-no-sc",
                  "--no-static-condensation", "Enable static condensation.");
   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
                  "--no-visualization",
                  "Enable or disable GLVis visualization.");
   args.Parse();
   if (!args.Good())
   {
      if (myid == 0)
      {
         args.PrintUsage(cout);
      }
      MPI_Finalize();
      return 1;
   }
   if (myid == 0)
   {
      args.PrintOptions(cout);
   }

   // 3. Read the (serial) mesh from the given mesh file on all processors.  We
   //    can handle triangular, quadrilateral, tetrahedral, hexahedral, surface
   //    and volume meshes with the same code.
   Mesh *mesh = new Mesh(mesh_file, 1, 1);
   int dim = mesh->Dimension();

   //mesh->UniformRefinement(); // this changes the failure to success

   // 5. Define a parallel mesh by a partitioning of the serial mesh. Refine
   //    this mesh further in parallel to increase the resolution. Once the
   //    parallel mesh is defined, the serial mesh can be deleted.
   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
   delete mesh;

   pmesh->UniformRefinement();

   pmesh->PrintInfo(std::cout);
   if (myid == 0)
       std::cout << "\n";

   Array<int> ess_bdrSigma(pmesh->bdr_attributes.Max());
   ess_bdrSigma = 0;
   ess_bdrSigma[0] = 1;

   int feorder = 0;
   //int dim = 3;
   FiniteElementCollection *hdiv_coll_test = new RT_FECollection(feorder, dim);
   FiniteElementCollection *l2_coll_test = new L2_FECollection(feorder, dim);
   FiniteElementCollection *hdivfree_coll_test = new ND_FECollection(feorder + 1, dim);

   ParFiniteElementSpace *R_space_test = new ParFiniteElementSpace(pmesh, hdiv_coll_test);
   ParFiniteElementSpace *W_space_test = new ParFiniteElementSpace(pmesh, l2_coll_test);
   ParFiniteElementSpace *C_space_test = new ParFiniteElementSpace(pmesh, hdivfree_coll_test);

   if (myid == 0)
    ess_bdrSigma.Print();

   ParDiscreteLinearOperator Divfree_op(C_space_test, R_space_test); // from Hcurl or HDivSkew(C_space) to Hdiv(R_space)
   Divfree_op.AddDomainInterpolator(new CurlInterpolator);
   Divfree_op.Assemble();
   Vector tempsol(Divfree_op.Width());
   tempsol = 0.0;
   Vector temprhs(Divfree_op.Height());
   temprhs = 0.0;
   Divfree_op.EliminateTrialDofs(ess_bdrSigma, tempsol, temprhs);
   //Divfree_op.EliminateTestDofs(ess_bdrSigma);
   Divfree_op.Finalize();
   HypreParMatrix * Divfree_hpmat = Divfree_op.ParallelAssemble();

   ParMixedBilinearForm *Bblock = new ParMixedBilinearForm(R_space_test, W_space_test);
   Bblock->AddDomainIntegrator(new VectorFEDivergenceIntegrator);
   Bblock->Assemble();
   //Vector tempsol(Bblock->Width());
   //tempsol = 0.0;
   //Vector temprhs(Bblock->Height());
   //temprhs = 0.0;
   //Bblock->EliminateTrialDofs(ess_bdrSigma, tempsol, temprhs);
   //Bblock->EliminateTestDofs(ess_bdrSigma);
   Bblock->Finalize();
   HypreParMatrix * Constraint_global = Bblock->ParallelAssemble();

   delete Bblock;

   HypreParMatrix * checkprod = ParMult(Constraint_global, Divfree_hpmat);

   SparseMatrix diagg;
   checkprod->GetDiag(diagg);

   SparseMatrix offdiagg;
   HYPRE_Int * cmap_offd;
   checkprod->GetOffd(offdiagg, cmap_offd);

   for (int i = 0 ;i < num_procs; ++i)
   {
       if (myid == i)
       {
           std::cout << "I am " << myid << "\n" << std::flush;
           std::cout << "Constraint[0] * Curl[0] diag norm = " << diagg.MaxNorm() << "\n";
           std::cout << "Constraint[0] * Curl[0] offdiag norm = " << offdiagg.MaxNorm() << "\n";
       }
       MPI_Barrier(comm);
   }

   delete checkprod;

   MPI_Finalize();
   return 0;
}
```
Thanks,
Kirill

**Comment by kvoronin** [comment]
Maybe, ParMesh::RefineGroups() or MarkTetMeshForRefinement() are to be called? Though, there is no mechanism for providing second input argument to the first routine (why is it declared as public?) and the second routine is protected so it cannot be called from the outside.

**Comment by mlstowell** [comment]
Hi, Kirill,

I'm trying to reproduce this issue (in MFEM's current master branch) but the norms always seem to be zero.  Perhaps if I used the same mesh I would see non-zeros.  Can you attach your mesh or let me know where I might find it.

Thanks,
Mark

**Comment by kvoronin** [comment]
Hi Mark,
Thanks for the response, the mesh file can be found at https://github.com/kvoronin/mfem/blob/dev_divfree_minsolver/data/cube_3d_moderate.mesh, please let me know if you are able to see the same strange behavior as I did.

**Comment by kvoronin** [comment]
Hi again, @mlstowell,
I've just checked with the current MFEM's master branch and I do see nonzero norm for 5 mpi processes (unlike 1-4 at least), with the given mesh.

**Comment by mlstowell** [comment]
Hi, Kirill,

I tried running your test code with the number of processors ranging from 1 to 8.  I saw offd norms equal to one on processors 1 and 2 for the five processor case.  All other numbers of processors showed norms of zero everywhere.

I must say, though, that you seem to be using MFEM in an unexpected manner.  I think of this as one of MFEM's dark corners.  The EliminateTrialDofs method is in the MixedBilinearForm class and ParDiscreteLinearOperator inherits from this class but technically it isn't a bilinear form because it does not compute an integral.  Also, the ParDiscreteLinearOperator::ParallelAssemble method does not perform any interprocess communication whereas ParMixedBilinearForm::ParallelAssemble does perform communication.

With this in mind I'm not surprised that EliminateTrialDofs does not produce the results you expect.  The question remains how do we form the operator that you need?   How do you intend to use a discrete curl operator which requires eliminating columns?

Thanks,
Mark

**Comment by kvoronin** [comment]
Thanks for the thoughts, @mlstowell!

What I really need is a Curl operator which acts onto Hdiv space with some part of the boundary Gamma said to be essential, and which Satisfies Div Curl = 0 and I want the operators to be assembled in some global form, which is MFEM's HypreParMatrix. The orthogonality property is used in the solver which must return a divergence-free solution and the last operation inside is the action of the Curl). 

Probably you see how this can be achieved in a different way?

Btw, for me it looks like a mesh handling problem. Isn't it strange that ParMesh calls the same Refinement() routine as Mesh, thus parallel structure is not taken into account? And the hanging function RefineGroups for which it's unclear how to get the second required argument makes me feel more sure about that. Can't anything of the just described be related to the observed problem?

Thanks,
Kirill

**Comment by mlstowell** [comment]
Hi, Kirill,

When you use EliminateTrialDofs on the Curl operator you're not enforcing the constraint that the vector field is zero on the essential boundary.  You're just ignoring whatever values the field might have on that portion of the boundary.  Typically such a boundary condition would be enforced on the field before the curl operator is used.  This not only sets the field to zero on the boundary but also ensures that the field approaches zero smoothly near the boundary.

The curl operator that you built in your test code is correct except for the call to EliminateTrialDofs.  You could also build a divergence operator, in a completely analogous manner, using the DivergenceInterpolator.  The product of these two operators should always be zero.

In other words if you take any field expressed as a ParGridFunction using the ND_FECollection, act on this field with the Curl operator you will get a field in HDiv (a ParGridFunction using the RT_FECollection).  If you then act on this HDiv field with the Divergence operator (built using the DivergenceInterpolator) you should get a result which is zero to machine precision.

The real question is, how should you compute the HCurl field so that it is zero on the boundary?  Aside from the boundary constraint what equation must this field satisfy on the interior of the problem domain?

Best wishes,
Mark

**Comment by tzanio** [comment]
Hey guys,

A quick comment that hopefully is helpful: if you split the degrees of freedom into interior "i" and boundary "b", then the discrete gradient matrix has the form
```
G = [G_ii G_ib]
    [  0  G_bb]
```
the `G_bi` block is zero, because boundary edges connect only boundary vertices (by definition).

Similarly, the discrete curl matrix is of the form
```
C = [C_ii C_ib]
    [  0  C_bb]
```
where again, boundary faces connect only boundary edges. (These are mentioned for example in this paper: http://epubs.siam.org/doi/abs/10.1137/110859361)

Now `C G = 0` holds in general, and `C G phi = 0` will certainly hold if `phi_b = 0`.  Moreover `phi_b = 0` implies `(G phi)_b = 0`, which in turn implies `(C G phi)_b = 0` so the zero Dirichlet values are "preserved" by the full `G` and `C` operators.

If you really need to eliminate the boundary dofs (which I'm not sure is ever necessary), than I suggest you delete the "b" columns in `G` and `C`, i.e. use
```
G' = [G_ii 0]
     [  0  0]
```
and
```
C' = [C_ii 0]
     [  0  0]
```
This way `C' G' = 0` continues to hold and boundary values are simply ignored.

**Comment by tzanio** [comment]
In the last case, you should probably use `ParFiniteElementSpace::GetEssentialTrueDofs` as in example 1. This makes sure that vertex and edge dofs on the Dirichlet boundary are marked as essential, even if the MPI task that owns them does not have boundary faces that contain them.

**Comment by kvoronin** [comment]
Thanks, @tzanio  and @mlstowell.
As far as I understand you, it is just invalid to eliminate anything from the curl operator. I needed that for forming a matrix Curl^T M Curl with some M acting in H(div, Gamma), and I wanted the resulting operator acting in H(curl, Gamma) for some essential boundary Gamma. 
In the end I decided that I can use the C^T M C without boundary conditions in H(curl) and manually correct the input and the resulting vectors at the boundary and be ok with that.

**Comment by mlstowell** [comment]
Hi, Kirill,

We often work with that same operator.  In fact we have two separate bilinear form integrators for computing that operator; CurlCurlIntegrator and MixedCurlCurlIntegrator.  The second is slightly more general and does a bit more error checking. You can look at examples ex3p and ex13p to see how they are used and how to apply boundary conditions.

Incidentally they are completely equivalent to Curl^T M Curl with M in H(div, Gamma).  I believe we may even have a unit test that verifies this equivalence.

Best wishes,
Mark

**Comment by kvoronin** [comment]
Hi @mlstowell,
Yeah, but my approach was to take M in the form of HypreParMatrix as an input (for some sort of the the solver), so I am not sure that it is possible to get a local matrix coefficient Q in this case.
Thanks,
Kirill
